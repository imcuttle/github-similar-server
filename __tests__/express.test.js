/**
 * @file main
 * @author imcuttle
 * @date 2018/4/4
 */
const express = require('express')
const nps = require('path')
const request = require('supertest')

const { makeFixture } = require('./helper')
const middleware = require('../src/index')

describe('main', function() {
  const app = express()

  app.use('/case-1', middleware({ root: makeFixture('case-1') }))
  app.use('/case-404', middleware({ root: makeFixture('case-404') }))
  app.use('/case-readme', middleware({ root: makeFixture('case-readme') }))
  app.use('/case-raw', middleware({ root: makeFixture('case-raw') }))
  app.use('/case-zh', middleware({ root: makeFixture('case-zh') }))
  app.use('/case-meta', middleware({ root: makeFixture('case-meta') }))
  app.use('/case-single-file', middleware({ root: makeFixture('single-file.md') }))

  app.use(
    '/case-template-string',
    middleware({
      root: makeFixture('case-zh'),
      markdownTemplateString: '<%=title%> <%=foo%>',
      templateParameters: {
        title: 'abc',
        foo: 'fooo'
      }
    })
  )

  it('should static works when path not found', function(done) {
    request(app)
      .get('/case-1/not_found')
      .expect(404)
      .end(function(err, res) {
        if (err) throw err
        done()
      })
  })

  it('should support meta title firstly', function(done) {
    request(app)
      .get('/case-meta/')
      .expect(200)
      .end(function(err, res) {
        if (err) throw err
        const dom = new DOMParser().parseFromString(res.text, 'text/html')
        expect(dom.title).toBe('metaTitle')
        expect(dom.body.innerHTML).toMatchSnapshot()
        done()
      })
  })

  it('should static works when path not found, but has 404.html', function(done) {
    request(app)
      .get('/case-404/not_found')
      .expect(function(res) {
        expect(res.text).toMatchSnapshot()
      })
      .expect(404)
      .end(done)
  })

  it('should return 302 when directory matched', function(done) {
    request(app)
      .get('/case-readme/readme')
      .expect(function(res) {
        expect(res.text).toMatchSnapshot()
      })
      .expect(302)
      .end(done)
  })

  it('should skip readme when index.html existed', function(done) {
    request(app)
      .get('/case-readme/')
      .expect(function(res) {
        expect(res.text).toBe('index\n')
      })
      .expect(200)
      .end(done)
  })

  it('should redirect to directory', function(done) {
    request(app)
      .get('/case-readme')
      .expect(302)
      .end(done)
  })

  it("should ecstatic's link is correct when has baseUrl", function(done) {
    request(app)
      .get('/case-1/')
      .expect(200)
      .expect(function(res) {
        expect(res.text).not.toContain('<a href="/abc/">')
        expect(res.text).toContain('<a href="/case-1/abc/">')
        expect(res.text).not.toContain('<a href="/haha.html">')
        expect(res.text).toContain('<a href="/case-1/haha.html">')
      })
      .end(done)
  })

  it('should responses raw markdown text when `?raw`', function(done) {
    request(app)
      .get('/case-raw/README.md?raw=true')
      .expect(function(res) {
        expect(res.text).toBe('# raw\n')
      })
      .expect(200)
      .end(function() {
        request(app)
          .get('/case-raw/readme/readme.md?raw=true')
          .expect(function(res) {
            expect(res.text).toBe('## hah\n')
          })
          .expect(200)
          .end(done)
      })
  })

  it('should response html when request path contains chinese', function(done) {
    request(app)
      .get(`/case-zh/${encodeURIComponent('你好 呀')}.md?a=22`)
      .expect(function(res) {
        expect(res.text).toMatchSnapshot()
      })
      .expect(200)
      .end(function() {
        request(app)
          .get(`/case-zh/${encodeURIComponent('你好 呀')}.md?raw=true`)
          .expect(function(res) {
            expect(res.text).toBe('## haha\n')
          })
          .expect(200)
          .end(done)
      })
  })

  it('should response html when through `markdownTemplateString` and `templateParameters`', function(done) {
    request(app)
      .get(`/case-template-string/${encodeURIComponent('你好 呀')}.md`)
      .expect(function(res) {
        expect(res.text).toEqual('你好 呀 fooo')
      })
      .expect(200)
      .end(done)
  })

  it('should works on single-file when root is extract file', function (done) {
    request(app)
      .get('/case-single-file')
      .expect(function(res) {
        expect(res.text).toContain('<p>single</p>\n')
      })
      .expect(200)
      .end(done)
  })

  it('should works on single-file when root is extract file 2', function (done) {
    request(app)
      .get('/case-single-file/')
      .expect(function(res) {
        expect(res.text).toContain('<p>single</p>\n')
      })
      .expect(200)
      .end(done)
  })

  it('should works on single-file when root is extract file (throw error)', function (done) {
    request(app)
      .get('/case-single-file/a')
      .expect(404)
      .end(done)
  })

  it('should works on single-file when root is extract file with query `?raw`', function (done) {
    request(app)
      .get('/case-single-file?raw=true')
      .expect(function(res) {
        expect(res.text).toEqual('single\n')
      })
      .expect(200)
      .end(done)
  })

  it('should githubSimilar.markdown', function () {
    expect(middleware.markdown).toBeInstanceOf(Function)
    expect(middleware.static).toBeInstanceOf(Function)
  })
})
