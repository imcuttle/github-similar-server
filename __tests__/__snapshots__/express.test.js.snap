// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`main should response html when request path contains chinese 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">
<head>
  <meta charset=\\"UTF-8\\">
  <meta http-equiv=\\"X-UA-Compatible\\" content=\\"IE=edge\\">
  <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0, user-scalable=no\\">
  <title>你好 呀</title>
  <link rel=\\"stylesheet\\" href=\\"//unpkg.com/github-markdown-css\\">
  <link rel=\\"stylesheet\\" href=\\"//unpkg.com/highlight.js/styles/github.css\\">
  <style>
    .markdown-body {
      max-width: 65%;
      margin: 40px auto 30px;
    }
    @media screen and (max-width: 1200px) {
      .markdown-body {
        max-width: 78%;
      }
    }

    @media screen and (max-width: 700px) {
      .markdown-body {
        max-width: 90%;
        margin: 20px auto 10px;
      }
    }
    @media screen and (max-width: 400px) {
      .markdown-body {
        max-width: 98%;
        margin: 20px auto 10px;
      }
    }
  </style>
</head>
<body>
<div class=\\"markdown-body\\">
  <h2 id=\\"haha\\"><a href=\\"#haha\\" class=\\"anchor\\"><svg aria-hidden=\\"true\\" class=\\"octicon octicon-link\\" height=\\"16\\" version=\\"1.1\\" view-box=\\"0 0 16 16\\" width=\\"16\\"><path d=\\"M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z\\"></path></svg></a>haha</h2>
</div>
</body>
</html>
"
`;

exports[`main should return 302 when directory matched 1`] = `"Found. Redirecting to /readme/"`;

exports[`main should static works when path not found, but has 404.html 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">
<head>
    <meta charset=\\"UTF-8\\">
    <title>404</title>
</head>
<body>

</body>
</html>
"
`;

exports[`main should support meta title firstly 1`] = `
"
<div class=\\"markdown-body\\">
  <h1 id=\\"raw\\"><a href=\\"#raw\\" class=\\"anchor\\"><svg aria-hidden=\\"true\\" class=\\"octicon octicon-link\\" height=\\"16\\" version=\\"1.1\\" view-box=\\"0 0 16 16\\" width=\\"16\\"><path d=\\"M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z\\"></path></svg></a>raw</h1>
</div>


"
`;
